### PR TITLE
Filter function based on metadata for Datasets

### LINKED ISSUE #256: Allow the data objects (module and dataset) to filter on contrast
Please:

- [x] Check for duplicate requests.
- [x] Describe your goal, and if possible provide a code snippet with a motivating example.

It would be nice to be able to do the following:

```python
data_module = FastMriDataModule(
    ...,
	test_contrasts='CORPD_FBK',
	val_contrasts='CORPD_FBK',
	train_contrasts=['CORPD_FBK', 'CORPDFS_FBK'],
)
```

This would enable:
- generalization tests
- finer evaluation

I want this to perform finer evaluation anyway, so I can do a PR if interested.

### PR DESCRIPTION
**Pull request by zaccharieramzi**
This should close #256

For the unit test I would like to modify the created temporary data, but don't know whether I can do that.

### CODE REVIEW

**Review by mmuckley — CHANGES_REQUESTED** [review]
Overall this looks good to me. The only thing is I'm wondering if we can tweak the API for the examples list. I made a suggestion based on named tuples that would allow strong typing for any example filter. Might also want to rename `self.examples` to something more meaningful like `self.samples`. If you're using an editor like PyCharm or VS Code you should be able to refactor the name without having to crawl through all the code.

Would you be willing to add that to this PR?

**Comment by zaccharieramzi** [comment]
@mmuckley Yes ofc always happy to help.

I am just wondering whether I should also apply the `example` -> `sample` change to the `banding_removal` subfolder? It has a lot of duplicated code atm, so not sure it's in a steady state anyway, no?

Also, I though that `raw_sample` was a bit more speaking than `sample`, because it allowed to make the difference with the sample classes defined in `transforms.py`. WDYT?

File: fastmri/data/mri_data.py [hunk 1]
```diff
@@ -270,7 +283,9 @@ def __init__(
                 metadata, num_slices = self._retrieve_metadata(fname)
 
                 self.examples += [
-                    (fname, slice_ind, metadata) for slice_ind in range(num_slices)
+                    (fname, slice_ind, metadata)
+                    for slice_ind in range(num_slices)
+                    if self.example_filter(metadata)
```
[the remark this answers is not in the document]
Comment (zaccharieramzi) [reply fastmri/data/mri_data.py]: All done, let me know what you think

File: tests/create_temp_data.py [hunk 2]
```diff
@@ -92,15 +92,19 @@ def create_temp_data(path):
...
+                    "padding_left": padding_left,
+                    "padding_right": padding_right,
+                    "encoding_size": enc_size,
+                    "recon_size": recon_size,
+                    "acquisition": acquisition,
+                }
+                if split.split("_")[-1] in ("test", "challenge"):
+                    metadata_dict["acceleration"] = 4
+
+                metadata[str(fname)] = (metadata_dict, num_slices)
```
[the remark this answers is not in the document]
Comment (zaccharieramzi) [reply tests/create_temp_data.py:107]: thanks !

**Review by mmuckley — APPROVED** [review]
This looks great to me and seems to be passing all tests.

Since this is an API change it could break a few things - I'm going to go a head and merge into `main` for now and closely monitor for issues - may require a follow-up PR.

Great work @zaccharieramzi!

**Pull request merged**